### PR TITLE
IDCOM-2089 Close Transaction Instruction

### PR DESCRIPTION
**Pull request by rado0x54**

### CODE REVIEW

File: packages/client/middleware/timeDelay/src/index.ts [hunk 1]
```diff
@@ -131,4 +131,9 @@ export class TimeDelayMiddleware
 
     return { instructions, signers: [] };
   }
+
+  public async onClose(): Promise<MiddlewareResult> {
+    // TODO: implement
```
Comment (JulianSpring) [inline packages/client/middleware/timeDelay/src/index.ts:136]: Does this need to be done?

File: programs/cryptid/src/instructions/close_transaction.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+cryptid_account_index: u32,
+/// The bump seed for the Did Account
+did_account_bump: u8,
+)]
+pub struct CloseTransaction<'info> {
+    /// The Cryptid instance to execute with
+    /// CHECK: Cryptid Account can be generative and non-generative
+    #[account(
+    mut,
+    // TODO(ticket): Verification done in instruction body. Move back with Anchor generator
```
Comment (JulianSpring) [inline programs/cryptid/src/instructions/close_transaction.rs:25]: should this  be done?

**Review by JulianSpring — COMMENTED** [review]
Just two todos that I am not sure if should be left in.

File: packages/client/middleware/timeDelay/src/index.ts (hunk 1 above)
Comment (rado0x54) [reply packages/client/middleware/timeDelay/src/index.ts:136]: Yes, but not tonight. It's not part of the audit-related program code. This specific middleware just needs to clean up it's state.

File: programs/cryptid/src/instructions/close_transaction.rs (hunk 2 above)
Comment (rado0x54) [reply programs/cryptid/src/instructions/close_transaction.rs:25]: There are 10x of the exact TODOs. It's a TODO that get's resolved when anchor provides the functionality.

File: packages/client/core/src/service/cryptid.ts [hunk 3]
```diff
@@ -198,8 +198,13 @@ export class CryptidService {
       cryptidAccountDetails: account,
     };
 
-    // Middlewares may have functions taht are executed on propose stage, on execute stage, or both
-    const middlewareFn = stage === "Propose" ? "onPropose" : "onExecute";
+    // Middlewares may have functions interface functions for propose, execute and close.
+    const middlewareFn =
```
Comment (tbarri) [inline packages/client/core/src/service/cryptid.ts:202]: Generally, I think we should avoid nested ternary statements as they are difficult to read. I would recommend moving this to a helper function with a switch statement. It also seems like one could just write:

 `middlewareFn = "on" + stage;`

**Review by tbarri — COMMENTED** [review]
For the future: I would really like to recommend splitting changes like this into even smaller commits. For example, the renaming of variables should be one commit. The adding of the auto-generated code should be another. The implementation of the client side should be one commit, and the rust side should be another. 

Splitting the commits in this way will really help reviewers to be able to see how you progressed through your changes, and to be able to filter out a lot of the refactoring noise. It's more effort on the implementer side, but it makes the world of difference for your reviewers. It improves the quality of the reviews as it allows reviewers to break it down into manageable components.

But otherwise, thank you for implementing this change!

**Pull request merged**